include/irr/asset/CTexturePacker.h

```diff
@@ -4,35 +4,96 @@
+            lin_addr_alctr_t::reserved_size(_squareSz*_squareSz, _squareSz*_squareSz, 1ull))
+        ),
+        m_addrAlctr(reinterpret_cast<void*>(m_alctrReservedSpace->data()), 0u, 0u, _squareSz*_squareSz, _squareSz*_squareSz, 1u)
+    {}
+
+    inline addr_type alloc_addr(size_t _subsquareSz)
+    {
+        const size_t cnt = _subsquareSz*_subsquareSz;
+        return m_addrAlctr.alloc_addr(cnt*cnt, cnt*cnt);
+    }
+
+    inline addr_type unpackAddress_x(addr_type _addr) const
+    {
+        return core::morton2d_decode_x(_addr&ADDR_MORTON_MASK);
+    }
+    inline addr_type unpackAddress_y(addr_type _addr) const
+    {
+        return core::morton2d_decode_y(_addr&ADDR_MORTON_MASK);
+    }
+    inline addr_type unpackAddress_layer(addr_type _addr) const
+    {
+        return (_addr>>ADDR_LAYER_SHIFT);
+    }
+};
+
 class CTexturePacker
 {
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t TILE_PADDING = 8u;
+
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_LAYER_SHIFT = 28u;//14bit 2d page address and 4bit for 3rd address dimension (layer)
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_MORTON_MASK = (1u<<ADDR_LAYER_SHIFT)-1u;
+
     using addr_alctr_t = core::GeneralpurposeAddressAllocator<uint32_t>;
 
+    static uint32_t megaimgLayerSize(uint32_t _pgSz, uint32_t _tilesPerDim)
+    {
+        return _tilesPerDim*(TILE_PADDING + _pgSz) + TILE_PADDING;
+    }
+    static core::vector2du32_SIMD pageCoords(uint32_t _x, uint32_t _y, uint32_t _pgSz)
+    {
+        core::vector2du32_SIMD coords(_x,_y);
+        coords *= (_pgSz + TILE_PADDING);
+        coords += TILE_PADDING;
+        return coords;
+    }
+
 public:
     struct STexOffset
     {
-        uint32_t x, y;
+        uint16_t x, y;
     };
 
-    CTexturePacker(E_FORMAT _format, uint32_t _szxy = 1u<<14, uint32_t _pgSzxy = 256u) :
+    CTexturePacker(E_FORMAT _format, uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u) :
         m_pgSzxy(_pgSzxy),
-        m_alctrReservedSpace(core::make_refctd_dynamic_array<decltype(m_alctrReservedSpace)>(
-            addr_alctr_t::reserved_size(_szxy / _pgSzxy, _szxy / _pgSzxy, 1ull))
-        ),
-        m_addrAlctr(reinterpret_cast<void*>(m_alctrReservedSpace->data()), 0u, 0u, _szxy/_pgSzxy, _szxy,_pgSzxy, 1u)
+        m_tilesPerDim(_tilesPerDim),
+        m_addrAlctr(14u, _tilesPerDim)
     {
-        assert(core::isPoT(_szxy));
+        assert(core::isPoT(_tilesPerDim));
         assert(core::isPoT(_pgSzxy));
 
+        const uint32_t megaimgSz = megaimgLayerSize(_pgSzxy, _tilesPerDim);
+
         ICPUImage::SCreationParams params;
-        params.extent = {_szxy,_szxy,1u};
+        params.extent = {megaimgSz,megaimgSz,1u};
         params.format = _format;
-        params.arrayLayers = 1u;
+        params.arrayLayers = _numLayers;
         params.mipLevels = 1u;
         params.type = IImage::ET_2D;
         params.samples = IImage::ESCF_1_BIT;
```

Comment:
split out this part, let there be an `ITexturePacker` and a `ICPUTexturePacker` (`IGPUTexturePacker` to come in the future)